lib/ProductOpener/Display.pm

```diff
@@ -10272,6 +10212,285 @@ CSS
+
+This function checks various conditions such as whether the nutrient has a value, is optional, or is country-specific, and returns a boolean indicating whether the nutrient should be shown.
+
+=head3 Arguments
+
+=head4 Nutrient $nutriment
+
+The nutrient identifier (e.g., "fat", "sodium", etc.).
+
+=head4 Product reference $product_ref
+
+Reference to the product's data, which contains the nutritional values.
+
+=head4 Request reference $request_ref
+
+Reference to request data, which can contain country-specific or other relevant information.
+
+=head3 Return values
+
+Returns 1 if the nutrient should be shown, 0 otherwise.
+
+=cut
+
+sub should_show_nutrient {
+	my ($nutriment, $product_ref, $request_ref, $nid) = @_;
+
+	# Logic to determine if the nutrient should be shown (as in your original code)
+	my $is_nutrient_with_value = (
+		((defined $product_ref->{nutriments}{$nid}) and ($product_ref->{nutriments}{$nid} ne ''))
+			or ((defined $product_ref->{nutriments}{$nid . "_100g"})
+			and ($product_ref->{nutriments}{$nid . "_100g"} ne ''))
+			or ((defined $product_ref->{nutriments}{$nid . "_prepared"})
+			and ($product_ref->{nutriments}{$nid . "_prepared"} ne ''))
+			or ((defined $product_ref->{nutriments}{$nid . "_modifier"})
+			and ($product_ref->{nutriments}{$nid . "_modifier"} eq '-'))
+			or ((defined $product_ref->{nutriments}{$nid . "_prepared_modifier"})
+			and ($product_ref->{nutriments}{$nid . "_prepared_modifier"} eq '-'))
+	);
+
+	# Logic to show important or non-optional nutrients
+	if (($nutriment !~ /-$/) or $is_nutrient_with_value) {
+		return 1;
+	}
+
+	# Hide non-important nutrients when no value is present
+	if ((($nutriment !~ /^!/) or ($product_ref->{id} eq 'search')) and not($is_nutrient_with_value)) {
+		return 0;
+	}
+
+	# Special handling for nutrition scores (optional, based on request_ref)
+	if ($nid =~ /^nutrition-score-(.*)$/) {
+		# Always show FR score and Nutri-Score
+		if (($request_ref->{cc} ne $1) and (not($1 eq 'fr'))) {
+			return 0;
+		}
+
+		# Experimental: hide nutrition scores for now
+		return 0;
+	}
+
+	return 1;
+}
+
+=head2 get_nutrient_level ( $nutriment )
+
+Determines the level of a nutrient based on the number of leading dashes.
+
+The level is used to establish whether a nutrient is a top-level nutrient or a sub-nutrient (e.g., fat -> saturated fat -> omega-3).
+
+=head3 Arguments
+
+=head4 Nutrient $nutriment
+
+The nutrient identifier string, which may contain leading dashes to denote levels.
+
+=head3 Return values
+
+Returns an integer indicating the nutrient's level.
+- 0 for main nutrients
+- 1 for sub-nutrients
+- 2 for sub-sub-nutrients, etc.
+
+=cut
+
+sub get_nutrient_level {
+	my ($nutriment) = @_;
+	if ($nutriment =~ /^!?--/) {
+		return 2;    # Sub-sub-nutrient (level 2)
+	}
+	elsif ($nutriment =~ /^!?-/) {
+		return 1;    # Sub-nutrient (level 1)
+	}
+	return 0;    # Top-level nutrient (level 0)
+}
```

Comment:
I am +1 for the refactorings in this branch; however I wonder whether it might make sense to apply those as a separate PR before the fix itself (the `add_parent_if_missing` step).

If I can understand it already, then perhaps someone else reading this later could too -- so maybe it would be unnecessary additional effort.

The other annoying detail that would add even more work, though, would be if I pointed out that `get_nutrient_level` and `should_show_nutrient` now seem to be perfect candidates for unit test coverage (small, self-contained logic, where breakage could be problematic).  Could we add some test coverage to them?